src/pipx/commands/upgrade.py

```diff
@@ -67,21 +68,20 @@ def upgrade(
     display_name = f"{package_metadata.package}{package_metadata.suffix}"
     new_version = package_metadata.package_version
 
-    expose_apps_globally(
+    # remove existing links, in case a binary is not provided in new version
+    unexpose_package_globally(constants.LOCAL_BIN_DIR, package_metadata)
```

Comment:
Technically unrelated to this PR. I realized that a package might not provide a binary anymore after an upgrade. If that happens, a dead symlink would remain. The fix is to remove everything and expose then binaries again.